File: Python Programs/unit5/MergeSort.py
from __future__ import print_function

# Merges two subarrays of arr[]


def merge(arr, low, mid, high):
    n1 = mid - low + 1
    n2 = high - mid
    # temp arrays
    leftSubArray = [0] * (n1)
    rightSubArray = [0] * (n2)

    for i in range(0, n1):
        leftSubArray[i] = arr[low + i]
    for j in range(0, n2):
        rightSubArray[j] = arr[mid + 1 + j]
    i = 0
    j = 0
    k = low
    while i < n1 and j < n2:
        if leftSubArray[i] <= rightSubArray[j]:
            arr[k] = leftSubArray[i]
            i += 1
        else:
            arr[k] = rightSubArray[j]
            j += 1
        k += 1
    # copy the remaining elements leftSubArray if any
    while i < n1:
        arr[k] = leftSubArray[i]
        i += 1
        k += 1
    # Copy the remaining elements of rightSubArray if any
    while j < n2:
        arr[k] = rightSubArray[j]
        j += 1
        k += 1


def mergeSort(arr, low, high):
    if low < high:

        mid = (low + high) / 2

        # Sort first and second halves
        mergeSort(arr, low, mid)
        mergeSort(arr, mid + 1, high)
        merge(arr, low, mid, high)



# replace raw_input with input if Python3
arr = [int(x) for x in raw_input(
    "Enter the elements of the array with space : \n").split(" ")]
n = len(arr)

mergeSort(arr, 0, n - 1)
print ("\n\nSorted array is")
print(arr)

File: Java Programs/Chapter 5/MergeSort/src/mergesort/MergeSort.java

package mergesort;

import java.util.Scanner;

public class MergeSort {

public static void Display(int A[], int n) // Helper Function to Display the
												// Array
	{
		System.out.println();
		for (int i = 0; i < n; i++) {
			System.out.print(A[i] + "\t");
		}
	}
public static void MergeSort(int A[], int low, int high) {
	int mid;
        System.out.println("\nlow = "+low+", high = "+high);
    	if (low < high)
        {
            mid = (high + low)/2;
            System.out.println("\nmid = "+mid+ ", Call Merge_Sort(A, "+low+","+mid+")");
            MergeSort(A, low, mid);
            System.out.println("\nmid = "+mid+ ", Call Merge_Sort(A, "+(mid+1)+","+high+")");
            MergeSort(A, mid + 1, high);
            System.out.println("\nmid = "+mid+", Call Merge(A,"+low+","+mid+","+high+")");
            Merge(A, low, mid, high);
            System.out.println("\nend Merge(A,"+low+","+mid+","+high+")");
        }
	//Display(A, n);
	}

	public static void Merge(int a[], int low, int mid, int high)
	{
	    int temp[]=new int[100];
	    int subarray1 = low;
	    int subarray2 = mid + 1;
	    int temparray = low;
	    int i;


	   	while (subarray1 <= mid && subarray2 <= high)
		{
	        if (a[subarray1] < a[subarray2])
	        {
	    		temp[temparray] = a[subarray1];
				subarray1++;
			}
	        else
			{
	    		temp[temparray] = a[subarray2];
				subarray2++;
			}
			temparray++;

	    }


		if(subarray1 > mid)
		{
		    while (subarray2 <= high)
	     	{
				temp[temparray++] = a[subarray2++];
	  		}
		}
		else
		{
			while (subarray1 <= mid)
			{
	        	temp[temparray++] = a[subarray1++];
	  		}
		}

	   	for(i=low;i<=high;i++)
		{
			a[i] = temp[i];
		}
	}

    public static void main(String[] args) {
        Scanner sc=new Scanner(System.in);
        System.out.println("Enter the no of elements :");
        int n = sc.nextInt();
		int A[] = new int[n];
                System.out.println("Enter the elements :");

		for (int i = 0; i < n; i++)
			A[i] = sc.nextInt();
                System.out.println("Input Array : ");	
		Display(A,n);
		MergeSort(A,0, n-1);
		Display(A,n);
    }
}
